### PR TITLE
DATACASS-335 - Add support for reactive data access.

### PR DESCRIPTION
**Pull request by mp911de**
We now support reactive data access with Spring Data Cassandra by adopting Datastax' asynchronous driver.

`ReactiveCqlTemplate` and `ReactiveCassandraTemplate` use [Project Reactor](http://projectreactor.io/) wrapper types `Mono` and `Flux` to implement Template API and repository support. Reactive template supports common operations such as:
- Query/Execution methods for static CQL and prepared statements
- Insert/Save/Update/Delete methods
- Exists and Count projections
- Reactive Callback methods

``` java
Person person = new Person("Dave", 25);

template.insert(person) //
    .flatMap(p -> template.update(new Person("Sven", 25))) //
    .flatMap(p -> template.selectOneById(person.getId(), Person.class)) //
    .subscribeWith(TestSubscriber.create()) //
    .await() //
    .assertValuesWith(result -> {
        assertThat(result.getFirstName(), is(equalTo("Sven")));
    });
```

Reactive Repository support is built on top of ReactiveCassandraTemplate using ReactiveCassandraRepository as the store-specific base repository. Reactive repositories are enabled by using`@EnableReactiveCassandraRepositories` on a `@Configuration` class to opt-in for reactive support. Reactive repositories can be composed of a reactive base interface such as
- `ReactiveCrudRepository`
- `ReactiveSortingRepository`
- `RxJava1CrudRepository`
- `RxJava1SortingRepository`

and are identified as reactive repository if one method uses a reactive wrapper type (such as `Flux` or `Observable`). If a reactive repository is discovered, it's not implemented by the blocking repository support but with the reactive repository factory. Blocking methods are not (yet) synchronized when using a reactive repository so each repository method must use a reactive wrapper result type. Reactive repository support with Spring Data allows using RxJava1 and Project Reactor types to declare repository methods. Reactive wrapper types are internally converted so the composition library choice on repository level is left up to the user.

There's feature parity between Reactive Cassandra repository support and blocking repository support.

Feature overview:
- Query Methods using String queries and Query Derivation
- Projections

``` java
@Configuration
@EnableReactiveCassandraRepositories
class ApplicationConfig extends AbstractReactiveCassandraConfiguration {

  @Override
  protected String getKeyspaceName() {
    return "mykeyspace";
  }

  @Override
  protected String getEntityBasePackages() {
    return new String[] {"com.springdata.cassandra"};
  }
}

public interface PersonRepository extends ReactiveSortingRepository<Person, String> {

  Flux<Person> findByFirstname(String firstname);

  Flux<Person> findByFirstname(Publisher<String> firstname);

  Mono<Person> findByFirstnameAndLastname(String firstname, String lastname);
}

public interface PersonRepository extends RxJava1SortingRepository<Person, String> {

  Observable<Person> findByFirstname(String firstname);

  Observable<Person> findByFirstname(Single<String> firstname);

  Single<Person> findByFirstnameAndLastname(String firstname, String lastname);
}
```

---

Related ticket: [DATACASS-335](https://jira.spring.io/browse/DATACASS-335)
Dependencies:
- [DATACMNS-836](https://jira.spring.io/browse/DATACMNS-836)
- https://github.com/spring-projects/spring-data-commons/pull/177

### CODE REVIEW

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/ReactiveCassandraOperations.java [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+	<T> Mono<T> insert(T entity, WriteOptions options) throws DataAccessException;
+
+	/**
+	 * Insert the given entities and emit the entity if the insert was applied.
+	 *
+	 * @param entities The entities to insert, must not be {@literal null}.
+	 * @return the inserted entities.
+	 * @throws DataAccessException if there is any problem issuing the execution.
+	 */
+	<T> Flux<T> insert(Iterable<? extends T> entities) throws DataAccessException;
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/ReactiveCassandraOperations.java]: As discussed with @jxblum: Remove those methods initially until there's real demand.

**Comment by jxblum** [comment]
Note DATACMNS-836 needs to be merged into _Spring Data Commons_ `master` for the _Spring Data Cassandra_ **2.0.x** branch to build.

**Comment by jxblum** [comment]
Note, to run the _Spring Data Cassandra_ Maven build, I merged in `DATACMNS-836` to the _Spring Data Commons_ **2.0.x** branch, set the `version` of the artifact to **2.0.0.BUILD-SNAPSHOT**, and ran a build to update my local Maven repo.

**Comment by jxblum** [comment]
Committed to `2.0.x` in revisions [52f4f57](https://github.com/spring-projects/spring-data-cassandra/commit/52f4f570a99be2dd823c0288038d5496f122b520), [e0741a8](https://github.com/spring-projects/spring-data-cassandra/commit/e0741a8691d3a674d49a3d87cf18d0035953d3a9) and [1c7413b](https://github.com/spring-projects/spring-data-cassandra/commit/1c7413bf1f4102220895680d8bcc18683802dcd2) for the _Spring Data Cassandra_ **2.0 M1 (Kay)** release

**Comment by mp911de** [comment]
Spring Data Commons Reactive changes are currently under review. We should expect further changes.

**Pull request closed without merging**